### PR TITLE
[improve][client] Adding support for token supplier for the AuthenticationToken

### PR DESCRIPTION
**Pull request by Bouk250**
Master Issue: #396

### Motivation

The `AuthenticationToken` only supports static token, limiting us to use token without expiration.
So the integration with existing authentication returns tokens with expiration becomes not possible.
 

### Modifications

Added the necessary cpp code for support token supplier as a js callback returns string tokens.
#### Modified file:
- `index.d.ts` - Update constructor signature for support TS.
- `src/Authentication.cc` - Implement the logic of communication between js callback and cpp token supplier.
#### New file
-`src/TokenSupplier.h` - Needed for the logic implementation.
-`examples/consummer_token.js` - Example
### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
- Added tow new test in the e2e file to cover token supplier (sync/async)

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [X] `doc-not-needed` 
The changes is internally only, the feature is already documented.

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Comment by Bouk250** [comment]
@shibd Hi, any idea why tests failed ? Its because the warning about the detectOpenHandles?

**Comment by Bouk250** [comment]
@shibd Hi :),
I fixed the issue with the test, I forget to release the callback. Now it will work.

File: tstest.ts [hunk 1]
```diff
@@ -66,18 +66,30 @@ import Pulsar = require('./index');
     scope: "scope"
   });
 
-  const authToken: Pulsar.AuthenticationToken = new Pulsar.AuthenticationToken({
+  const authToken1: Pulsar.AuthenticationToken = new Pulsar.AuthenticationToken({
     token: 'foobar',
   });
 
+  const authToken2: Pulsar.AuthenticationToken = new Pulsar.AuthenticationToken({
```
Comment (shibd) [inline tstest.ts:73]: Seems authToken2 and authToken3 are not used. We can remove it

File: examples/consummer_token_auth.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+const { exit } = require('process');
+const Pulsar = require('../');
+console.log("Starting consumer");
+(async () => {
+
+  const auth = new Pulsar.AuthenticationToken({token: async () => {
```
Comment (shibd) [inline examples/consummer_token_auth.js:25]: Can you help add an example for `token supplier`?

**Review by shibd — COMMENTED** [review]
Great work! Left some small comments.

File: tstest.ts (hunk 1 above)
Comment (Bouk250) [reply tstest.ts]: You right, I keep the async one.

File: examples/consummer_token_auth.js (hunk 2 above)
Comment (Bouk250) [reply examples/consummer_token_auth.js]: Sure !
I added a new example for simple token auth (what was already implemented).
I improved the example of the token supplier to be more clear and understandable.

**Comment by klaatu01** [comment]
Great stuff, I'd like this feature released in 1.13.0 soon if possible!

**Pull request merged**